# ‘sentiment_Analysis’ 

In [101]:
import pandas as pd
import numpy as np

In [102]:
sentiment_data= pd.read_csv(r'Sentiment.csv',header=0,index_col=0)
sentiment_data.head()

candidate  candidate_confidence relevant_yn  \
id                                                             
1   No candidate mentioned                   1.0         yes   
2             Scott Walker                   1.0         yes   
3   No candidate mentioned                   1.0         yes   
4   No candidate mentioned                   1.0         yes   
5             Donald Trump                   1.0         yes   

    relevant_yn_confidence sentiment  sentiment_confidence     subject_matter  \
id                                                                              
1                      1.0   Neutral                0.6578  None of the above   
2                      1.0  Positive                0.6333  None of the above   
3                      1.0   Neutral                0.6629  None of the above   
4                      1.0  Positive                1.0000  None of the above   
5                      1.0  Positive                0.7045  None of the above   

    subject_matter_confidence candidate_gold             name  \
id                                                              
1                      1.0000            NaN       I_Am_Kenzi   
2                      1.0000            NaN    PeacefulQuest   
3                      0.6629            NaN    PussssyCroook   
4                      0.7039            NaN  MattFromTexas31   
5                      1.0000            NaN       sharonDay5   

   relevant_yn_gold  retweet_count sentiment_gold subject_matter_gold  \
id                                                                      
1               NaN              5            NaN                 NaN   
2               NaN             26            NaN                 NaN   
3               NaN             27            NaN                 NaN   
4               NaN            138            NaN                 NaN   
5               NaN            156            NaN                 NaN   

                                                 text tweet_coord  \
id                                                                  
1   RT @NancyLeeGrahn: How did everyone feel about...         NaN   
2   RT @ScottWalker: Didn't catch the full #GOPdeb...         NaN   
3   RT @TJMShow: No mention of Tamir Rice and the ...         NaN   
4   RT @RobGeorge: That Carly Fiorina is trending ...         NaN   
5   RT @DanScavino: #GOPDebate w/ @realDonaldTrump...         NaN   

                tweet_created            tweet_id tweet_location  \
id                                                                 
1   2015-08-07 09:54:46 -0700  629697200650592256            NaN   
2   2015-08-07 09:54:46 -0700  629697199560069120            NaN   
3   2015-08-07 09:54:46 -0700  629697199312482304            NaN   
4   2015-08-07 09:54:45 -0700  629697197118861312          Texas   
5   2015-08-07 09:54:45 -0700  629697196967903232            NaN   

                 user_timezone  
id                              
1                        Quito  
2                          NaN  
3                          NaN  
4   Central Time (US & Canada)  
5                      Arizona

In [103]:
sentiment_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13871 entries, 1 to 13871
Data columns (total 20 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   candidate                  13775 non-null  object 
 1   candidate_confidence       13871 non-null  float64
 2   relevant_yn                13871 non-null  object 
 3   relevant_yn_confidence     13871 non-null  float64
 4   sentiment                  13871 non-null  object 
 5   sentiment_confidence       13871 non-null  float64
 6   subject_matter             13545 non-null  object 
 7   subject_matter_confidence  13871 non-null  float64
 8   candidate_gold             28 non-null     object 
 9   name                       13871 non-null  object 
 10  relevant_yn_gold           32 non-null     object 
 11  retweet_count              13871 non-null  int64  
 12  sentiment_gold             15 non-null     object 
 13  subject_matter_gold        18 non-null     obj

In [105]:
sentiment_data.describe()

candidate_confidence  relevant_yn_confidence  sentiment_confidence  \
count          13871.000000            13871.000000          13871.000000   
mean               0.855689                0.927304              0.756936   
std                0.241388                0.141696              0.217682   
min                0.222200                0.333300              0.186000   
25%                0.674200                1.000000              0.651700   
50%                1.000000                1.000000              0.681300   
75%                1.000000                1.000000              1.000000   
max                1.000000                1.000000              1.000000   

       subject_matter_confidence  retweet_count      tweet_id  
count               13871.000000   13871.000000  1.387100e+04  
mean                    0.782801      45.803331  6.296058e+17  
std                     0.258215     153.981724  9.611863e+13  
min                     0.222200       0.000000  6.294531e+17  
25%                     0.641300       0.000000  6.294861e+17  
50%                     1.000000       2.000000  6.296726e+17  
75%                     1.000000      44.000000  6.296882e+17  
max                     1.000000    4965.000000  6.297017e+17

In [104]:
sentiment_data.shape

(13871, 20)

In [106]:
sentiment_data.isnull().sum()

candidate                       96
candidate_confidence             0
relevant_yn                      0
relevant_yn_confidence           0
sentiment                        0
sentiment_confidence             0
subject_matter                 326
subject_matter_confidence        0
candidate_gold               13843
name                             0
relevant_yn_gold             13839
retweet_count                    0
sentiment_gold               13856
subject_matter_gold          13853
text                             0
tweet_coord                  13850
tweet_created                    0
tweet_id                         0
tweet_location                3912
user_timezone                 4403
dtype: int64

In [107]:
sentiment_data.text.head(1)

id
1    RT @NancyLeeGrahn: How did everyone feel about...
Name: text, dtype: object

## take only text and sentiment column

In [108]:
sentiment_data=sentiment_data[['text','sentiment']]
sentiment_data

text sentiment
id                                                                
1      RT @NancyLeeGrahn: How did everyone feel about...   Neutral
2      RT @ScottWalker: Didn't catch the full #GOPdeb...  Positive
3      RT @TJMShow: No mention of Tamir Rice and the ...   Neutral
4      RT @RobGeorge: That Carly Fiorina is trending ...  Positive
5      RT @DanScavino: #GOPDebate w/ @realDonaldTrump...  Positive
...                                                  ...       ...
13867  RT @cappy_yarbrough: Love to see men who will ...  Negative
13868  RT @georgehenryw: Who thought Huckabee exceede...  Positive
13869  RT @Lrihendry: #TedCruz As President, I will a...  Positive
13870  RT @JRehling: #GOPDebate Donald Trump says tha...  Negative
13871  RT @Lrihendry: #TedCruz headed into the Presid...  Positive

[13871 rows x 2 columns]

In [109]:
new_data = sentiment_data[sentiment_data['sentiment'] != 'Neutral']

In [110]:
new_data.head()

text sentiment
id                                                             
2   RT @ScottWalker: Didn't catch the full #GOPdeb...  Positive
4   RT @RobGeorge: That Carly Fiorina is trending ...  Positive
5   RT @DanScavino: #GOPDebate w/ @realDonaldTrump...  Positive
6   RT @GregAbbott_TX: @TedCruz: "On my first day ...  Positive
7   RT @warriorwoman91: I liked her and was happy ...  Negative

In [111]:
new_data.sentiment.unique()

array(['Positive', 'Negative'], dtype=object)

# Print the total number of positive and negative sentiments

In [37]:
new_data.sentiment.value_counts()

Negative    8493
Positive    2236
Name: sentiment, dtype: int64

### Total number of positive sentiments = 2236

### Total number of positive sentiments = 8493

# Build a sequential RNN model to predict positive and negative sentiments.

In [112]:
from nltk.corpus import stopwords   # to get collection of stopwords
from sklearn.model_selection import train_test_split       # for splitting dataset
from tensorflow.keras.preprocessing.text import Tokenizer  # to encode text to int
from tensorflow.keras.preprocessing.sequence import pad_sequences   # to do padding or truncating
from tensorflow.keras.models import Sequential     # the model
from tensorflow.keras.layers import Embedding, LSTM, Dense # layers of the architecture
from tensorflow.keras.callbacks import ModelCheckpoint   # save model
from tensorflow.keras.models import load_model   # load saved model
import re

In [113]:
import nltk
nltk.download('stopwords')

[nltk_data] Error loading stopwords: <urlopen error [WinError 10060] A
[nltk_data]     connection attempt failed because the connected party
[nltk_data]     did not properly respond after a period of time, or
[nltk_data]     established connection failed because connected host
[nltk_data]     has failed to respond>


False

In [114]:
english_stops = set(stopwords.words('english'))

In [115]:
x_data = new_data['text']       # Reviews/Input
y_data = new_data['sentiment']    # Sentiment/Output


In [116]:
# PRE-PROCESS REVIEW
x_data = x_data.replace({'<.*?>': ''}, regex = True)          # remove html tag
x_data = x_data.replace({'[^A-Za-z]': ' '}, regex = True)     # remove non alphabet
x_data = x_data.apply(lambda review: [w for w in review.split() if w not in english_stops])  # remove stop words
x_data = x_data.apply(lambda review: [w.lower() for w in review])   # lower case
    

In [117]:
# ENCODE SENTIMENT -> 0 & 1
y_data = y_data.replace('positive', 1)
y_data = y_data.replace('negative', 0)


In [118]:
print('text')
print(x_data, '\n')
print('Sentiment')
print(y_data)

text
id
2        [rt, scottwalker, didn, catch, full, gopdebate...
4        [rt, robgeorge, that, carly, fiorina, trending...
5        [rt, danscavino, gopdebate, w, realdonaldtrump...
6        [rt, gregabbott, tx, tedcruz, on, first, day, ...
7        [rt, warriorwoman, i, liked, happy, i, heard, ...
                               ...                        
13867    [rt, cappy, yarbrough, love, see, men, never, ...
13868    [rt, georgehenryw, who, thought, huckabee, exc...
13869    [rt, lrihendry, tedcruz, as, president, i, alw...
13870    [rt, jrehling, gopdebate, donald, trump, says,...
13871    [rt, lrihendry, tedcruz, headed, presidential,...
Name: text, Length: 10729, dtype: object 

Sentiment
id
2        Positive
4        Positive
5        Positive
6        Positive
7        Negative
           ...   
13867    Negative
13868    Positive
13869    Positive
13870    Negative
13871    Positive
Name: sentiment, Length: 10729, dtype: object


## Split Dataset

In [119]:
x_train, x_test, y_train, y_test = train_test_split(x_data, y_data, test_size = 0.2)

print('Train Set')
print(x_train, '\n')
print(x_test, '\n')
print('Test Set')
print(y_train, '\n')
print(y_test)

Train Set
id
6346              [gopdebate, ahem, http, co, hd, ukbursz]
7093     [realdonaldtrump, whine, whine, whine, boo, ho...
4711     [i, not, of, those, conservatives, who, suppor...
2920     [gopdebate, maybe, right, try, amp, destroy, o...
12550    [rt, ericstonestreet, trump, cam, hands, gopde...
                               ...                        
2698     [rt, steveamiri, how, is, jeff, dunham, workin...
4133     [rt, amymek, thank, tedcruz, calling, rhino, g...
1062     [rt, deborahpeasley, multiple, shots, scott, w...
8429     [rt, larryelder, trump, said, megyn, ask, nine...
11642             [rand, paul, vote, gopdebates, randpaul]
Name: text, Length: 8583, dtype: object 

id
6825     [topics, left, gopdebate, on, th, anniversary,...
9815     [rt, wilberforce, one, favorite, things, gopde...
12939    [trump, hand, gestures, second, season, one, p...
11655    [rt, rwsurfergirl, ask, trump, legitimate, que...
5339     [good, luck, everyone, competing, gopdebate, b..

In [120]:
def get_max_length():
    text_length = []
    for text in x_train:
        text_length.append(len(text))
    text_length = np.array(text_length)
    text_length = text_length[~np.isnan(text_length)]  # Exclude NaN values
    return int(np.ceil(np.mean(text_length)))

In [121]:
# ENCODE REVIEW
token = Tokenizer(lower=False)    # no need lower, because already lowered the data in load_data()
token.fit_on_texts(x_train)
x_train = token.texts_to_sequences(x_train)
x_test = token.texts_to_sequences(x_test)

max_length = get_max_length()

x_train = pad_sequences(x_train, maxlen=max_length, padding='post', truncating='post')
x_test = pad_sequences(x_test, maxlen=max_length, padding='post', truncating='post')

total_words = len(token.word_index) + 1   # add 1 because of 0 padding

print('Encoded X Train\n', x_train, '\n')
print('Encoded X Test\n', x_test, '\n')
print('Maximum review length: ', max_length)

Encoded X Train
 [[    1  5484     6 ...     0     0     0]
 [   10  1775  1775 ...     1     0     0]
 [    4    71   417 ...    59    42    11]
 ...
 [    2 13424  1589 ...    28   106  1834]
 [    2   421     8 ...   249     4   196]
 [  156   127   144 ...     0     0     0]] 

Encoded X Test
 [[1016  598    1 ...  159  595  764]
 [   2  887   28 ...  293  123  888]
 [   8  870  703 ...    0    0    0]
 ...
 [   2   22   14 ...  856   73    0]
 [   2  101 1798 ...    0    0    0]
 [2399    1 1497 ...   11   64    0]] 

Maximum review length:  13


In [122]:
# Define the embedding dimension
embedding_dim = 100
num_classes = 2 

In [123]:
from keras.models import Sequential
from keras.layers import Embedding, SimpleRNN, Dense

# Define the RNN model
model = Sequential()

# Add an Embedding layer
model.add(Embedding(input_dim=total_words, output_dim=embedding_dim, input_length=max_length))

# Add a SimpleRNN layer
model.add(SimpleRNN(units=128))

# Add a Dense layer for classification
model.add(Dense(units=num_classes, activation='softmax'))

# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Print the model summary
model.summary()

Model: "sequential_8"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_6 (Embedding)     (None, 13, 100)           1342500   
                                                                 
 simple_rnn_5 (SimpleRNN)    (None, 128)               29312     
                                                                 
 dense_5 (Dense)             (None, 2)                 258       
                                                                 
Total params: 1,372,070
Trainable params: 1,372,070
Non-trainable params: 0
_________________________________________________________________


Training
For training, it is simple. We only need to fit our x_train (input) and y_train (output/label) data. For this training, I use a mini-batch learning method with a batch_size of 128 and 5 epochs.

Also, I added a callback called checkpoint to save the model locally for every epoch if its accuracy improved from the previous epoch.

In [124]:
checkpoint = ModelCheckpoint(
    'models/RNN',
    monitor='accuracy',
    save_best_only=True,
    verbose=1
)

In [125]:
from sklearn.preprocessing import LabelEncoder
from keras.utils import to_categorical

# Convert string labels to integers
label_encoder = LabelEncoder()
y_train_encoded = label_encoder.fit_transform(y_train)
y_test_encoded = label_encoder.transform(y_test)

# Convert the target data to one-hot encoding
y_train_one_hot = to_categorical(y_train_encoded)
y_test_one_hot = to_categorical(y_test_encoded)

# Define the RNN model
model = Sequential()
model.add(Embedding(input_dim=total_words, output_dim=embedding_dim, input_length=max_length))
model.add(SimpleRNN(units=128))
model.add(Dense(units=num_classes, activation='softmax'))
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Train the model
model.fit(x_train, y_train_one_hot, epochs=5, batch_size=32 ,callbacks=[checkpoint])

Epoch 1/5
269/269 [==============================] - ETA: 0s - loss: 0.4062 - accuracy: 0.8284
Epoch 1: accuracy improved from -inf to 0.82838, saving model to models\RNN
INFO:tensorflow:Assets written to: models\RNN\assets
269/269 [==============================] - 14s 45ms/step - loss: 0.4062 - accuracy: 0.8284
Epoch 2/5
268/269 [============================>.] - ETA: 0s - loss: 0.1991 - accuracy: 0.9257
Epoch 2: accuracy improved from 0.82838 to 0.92567, saving model to models\RNN
INFO:tensorflow:Assets written to: models\RNN\assets
269/269 [==============================] - 13s 49ms/step - loss: 0.1992 - accuracy: 0.9257
Epoch 3/5
267/269 [============================>.] - ETA: 0s - loss: 0.1078 - accuracy: 0.9635
Epoch 3: accuracy improved from 0.92567 to 0.96318, saving model to models\RNN
INFO:tensorflow:Assets written to: models\RNN\assets
269/269 [==============================] - 12s 45ms/step - loss: 0.1081 - accuracy: 0.9632
Epoch 4/5
268/269 [============================>.

In [126]:
# Evaluate the model on the test data
loss, accuracy = model.evaluate(x_test, y_test_one_hot)

print("Test Loss:", loss)
print("Test Accuracy:", accuracy)

68/68 [==============================] - 1s 6ms/step - loss: 0.4207 - accuracy: 0.8509
Test Loss: 0.42068415880203247
Test Accuracy: 0.8508853912353516


In [127]:
y_pred = (model.predict(x_test)>0.5).astype("int32")

true = 0
for i, y in enumerate(y_test_one_hot):
    if np.array_equal(y, y_pred[i]):
        true += 1
        
print('Correct Prediction: {}'.format(true))
print('Wrong Prediction: {}'.format(len(y_pred) - true))
print('Accuracy: {}'.format(true/len(y_pred)*100))

68/68 [==============================] - 1s 5ms/step
Correct Prediction: 1826
Wrong Prediction: 320
Accuracy: 85.08853681267475


### Load Saved Model
##### Load saved model and use it to predict a movie review statement's sentiment (positive or negative).

In [128]:
loaded_model = load_model('models/RNN')

## Based on the model, check the sentiment for the following two sentences
#### a. 'He is a great leader.'
#### b. 'He is a terrible leader.'

In [132]:
# Preprocess the sentences
sentences = ['He is a great leader.', 'He is a terrible leader.']
sequences = token.texts_to_sequences(sentences)
tokenize_words = pad_sequences(sequences, maxlen=max_length, padding='post', truncating='post')
print(tokenize_words)


# Make predictions using the loaded model
predictions = loaded_model.predict(tokenize_words)

# Get the predicted sentiment labels
sentiments = [ 'Negative','Positive']
predicted_labels = [sentiments[np.argmax(pred)] for pred in predictions]

# Print the predicted sentiment for each sentence
for i, sentence in enumerate(sentences):
    print("Sentence: {}".format(sentence))
    print("Predicted Sentiment: {}".format(predicted_labels[i]))
    print()

[[ 50  60  97 267   0   0   0   0   0   0   0   0   0]
 [ 50  60 901 267   0   0   0   0   0   0   0   0   0]]
1/1 [==============================] - 0s 50ms/step
Sentence: He is a great leader.
Predicted Sentiment: Positive

Sentence: He is a terrible leader.
Predicted Sentiment: Negative

